bin/util/transifex.js

```diff
@@ -0,0 +1,574 @@
+    if (value == null) return;
+    if (value instanceof PluralForms)
+      this._translated[key] = PluralForms.empty(value.length);
+    else
+      this.get(key).clear();
+  }
+
+  clear() {
+    for (const key of Object.keys(this._translated))
+      this.delete(key);
+  }
+
+  // Visit each translation.
+  walk(callbacks) {
+    if (typeof callbacks === 'function') {
+      this.walk([callbacks]);
+      return;
+    }
+
+    for (const key of Object.keys(this._source)) {
+      const value = this.get(key);
+      if (value instanceof Translation) {
+        for (const callback of callbacks)
+          callback(key, value);
+      } else {
+        value.walk(callbacks);
+      }
+    }
+  }
+
+  toJSON(key) {
+    const keyIsIndex = /^\d+$/.test(key);
+
+    if (this.has('0')) {
+      const result = [];
+      let emptyPluralForms = 0;
+      let emptyObjects = 0;
+      for (let i = 0; i < this.size; i += 1) {
+        const k = i.toString();
+        if (!this.has(k))
+          logThenThrow(this, 'error converting object to array');
+        const value = this.get(k).toJSON(k);
+        result.push(value);
+        if (value === undefined)
+          emptyPluralForms += 1;
+        else if (Object.keys(value).length === 0)
+          emptyObjects += 1;
+      }
+      if (emptyPluralForms + emptyObjects === this.size) {
+        // `this` has only empty translations. If possible, we return
+        // `undefined` so that there is not an empty array in the JSON. However,
+        // we do not return `undefined` if doing so would result in a sparse
+        // array: JSON does not support sparse arrays.
+        return keyIsIndex ? [] : undefined;
+      }
+      if (emptyPluralForms !== 0) logThenThrow(this, 'sparse array');
+      return result;
+    }
+
+    const result = {};
+    for (const k of Object.keys(this._translated)) {
+      const value = this.get(k).toJSON(k);
+      if (value != null) result[k] = value;
+    }
+    return Object.keys(result).length !== 0 || keyIsIndex ? result : undefined;
+  }
+}
+
+// This will not work for a linked locale message in an i18n custom block, but
+// we do not use those yet.
+const copyLinkedLocaleMessage = (key, { source, parent }) => {
+  const path = pathOfLinkedMessage(source);
+  if (path == null) return;
+  let linked = parent.root;
+  for (let i = 0; linked != null && i < path.length; i += 1)
+    linked = linked.get(path[i]);
+  if (linked != null && linked.translated != null) parent.set(key, source);
+};
+
+const verifyDestructure = (_, { source, translated }) => {
+  if (translated == null || !equals(Array.from(source), Array.from(translated)))
+    logThenThrow({ source, translated }, 'mismatch for source locale');
+};
+
+// If a component interpolation is only partially translated, we remove the
+// partial translation so that the resulting text is not a mix of locales. We
+// also remove an array that is missing a translation, because JSON does not
```

Comment:
What does "an array that is missing a translation" concretely refer to? I'm not seeing where there are arrays involved.

Reply:
Let's say the source JSON looks like this:

```
{
  "introduction": [
    "one",
    "two",
    "three"
  ]
}
```

and "one" and "three" are translated, but "two" is not. In the translation JSON, it's not clear what the second element of the array should be. (It has to be something, because JSON does not support sparse arrays.)

```
{
  "introduction": [
    "ūnus",
    ?,
    "trēs"
  ]
}
```

It doesn't work to use `""` (https://github.com/kazupon/vue-i18n/issues/563). `null` may work, but Vue I18n logs a warning to the console, and I think this workaround might not be available in the next major version (which has some related changes). For now, we simply discard the entire array.

We're able to work around this if the empty array element is an object or another array. (If the second element were a missing component interpolation, we could use `{}`.) However, if an array directly contains an untranslated string, we discard the entire array.

Reply:
I see. What I find confusing is that this is part of `destructure` and the source Transifex file for that doesn't have any arrays. I see now that you consider the indexed members the arrays on the structured JSON side. A tiny bit of context around that would really help someone who comes in without deep experience with the different JSON formats involved. Something like "In vue-i18n, strings can be grouped in an array. Those are displayed separately in Transifex and may not all be translated. If any string in an array is missing in a translation, we discard the whole array."